### ISSUE TITLE
Clarify toolbar exception in Understanding Reflow 1.4.10

### ISSUE THREAD

**Issue by smhigley** [w3c/wcag#916]
We've been having a lively discussion within some teams around Microsoft about the toolbar wording in Understanding Reflow's content exceptions, and what it means. It has mostly coalesced into two separate questions:

1. What is a toolbar? Is it only an element with `role="toolbar"`, or anything that visually or functionally resembles a toolbar (e.g. any set of horizontal buttons that control nearby content)?

2. What is allowed to have a horizontal scrollbar when a toolbar is present? Is it the toolbar itself, or the toolbar along with the content it controls, or the entire UI of any site or app that contains a toolbar somewhere within it?

Part of the discussion comes from confusion around when a toolbar would require two-dimensional layout to preserve meaning or functionality. It's usually pretty simple to collapse overflowing toolbar actions into a "more actions" type menu without loosing meaning or function, and it's been hard to think of a practical example where this isn't the case (at least in my mental model of a toolbar). I think a practical example of a toolbar interface where reflow isn't possible could help clarify when to allow horizontal scroll, and when to reflow the toolbar actions.

For example, I imagine something simple like a rich text editor like the github comment textbox (or even the [toolbar example in APG](https://w3c.github.io/aria-practices/examples/toolbar/toolbar.html) should not be an exception since reflow is possible without loosing meaning or function.

Anecdotally, the interfaces that have had the hardest time with the reflow requirement have not had trouble with toolbars, but with multiple panes of related content or interactivity -- e.g. an interface similar to Visual Studio or Photoshop where concurrent access to multiple related panels is fairly central to most users. It might possible to collapse everything but a single panel at any given time, but it's not necessarily a wonderful user experience. It's hard to draw the line on when or if two dimensional layout is necessary for usage in these cases though, since it's a slope of UI complexity rather than a single type of control. It would be great to have guidance on these types of interfaces as well, and whether they would ever fall under the two-dimensional
layout exception.

**Comment by WayneEDick** [comment]
Hi Sara,

The most important part of the toolbar exemption is that it is for the
toolbar only.
One of the most common ways people with low vision are dragged into needing
horizontal scrolling is when a toolbar drags dependent text off the page.
That is prohibited by the SC.
I don't think an actual specification of "toolbar" is necessary, though it
might help in legal arguments.

If you don't drag text controlled by the toolbar off the page, sane readers
with low vision will be thrilled.

Not required, but very usable: If it is ambiguous when a toolbar is
truncated by the right border give the user an indicator that there is more
content to see. I have missed many a control not thinking it is there.

Best, Wayne

**Comment by JAWS-test** [comment]
> What is a toolbar? Is it only an element with role="toolbar", or anything that visually or functionally resembles a toolbar (e.g. any set of horizontal buttons that control nearby content)?

A toolbar does not need an ARIA role to be considered a toolbar according to this SC.

> What is allowed to have a horizontal scrollbar when a toolbar is present? Is it the toolbar itself, or the toolbar along with the content it controls, or the entire UI of any site or app that contains a toolbar somewhere within it?

If the content controlled by the toolbar falls under the exception itself and can be scrolled (e.g. an image in Photoshop), then one would only have to consider whether it is better to scroll the image and toolbar separately or together. On the one hand, many scrollbars can make the page confusing. On the other hand, separate scrolling can make work much easier.
If the remaining content is text (as in Word, for example), the toolbar should be wrapable or scrollable. The text must not be scrolled.

> Part of the discussion comes from confusion around when a toolbar would require two-dimensional layout to preserve meaning or functionality. It's usually pretty simple to collapse overflowing toolbar actions into a "more actions" type menu without loosing meaning or function,

If the actual buttons of the toolbar are no longer visible, but have to be shown first, then this is a limitation of function and meaning. In this case it can be better if the toolbar is scrollable. Scrolling is faster than showing the elements individually.

**Comment by smhigley** [comment]
@WayneEDick and @alastc thanks for the clarification! It's nice to know both that "toolbar" is a fairly permissive term, and that allowed overflow is generally confined to the toolbar itself. The presentation creator app example is also really helpful -- we've actually discussed how reflow would apply to exactly that type of app 😄.

Would there be any possibility of getting some of this into the Understanding doc? As much as I'd love to point teams to this issue, I think the Understanding page would have a bit more authority :).

If I were to choose specific parts to update, changing this line:
> This Success Criterion therefore does not apply to interfaces which provide toolbars. 

to clarify that the SC only allows a second scrollbar within the toolbar itself + adding a presentation creator example to the Examples section would be first on my wishlist, and clear up a lot of the discussions we've had about this.

Thanks again! This has been an interesting SC to dive into, particularly on the very app-y web apps.

**Comment by mraccess77** [comment]
If I have a menu or toolbar (such as those found in a desktop application) does that menu or toolbar have to meet this requirement?  Does it fall under and exception or not?  It's not clear if the toolbar or menu bar needs to collapse or not.

**Comment by scottaohara** [comment]
closing this issue as resolved per the updates made in #4055.  I'm sure there's further examples/techniques that could be made on this topic - but I'd suggest we open a new issue for those as needed.
